src/VisualStudio/CodeLens/ReferenceCodeLensProvider.cs

```diff
@@ -0,0 +1,293 @@
+            {
+                // we always get data through VS rather than Roslyn OOP directly since we want final data rather than
+                // raw data from Roslyn OOP such as razor find all reference results
+                var referenceCount = await _callbackService.InvokeAsync<ReferenceCount>(
+                    _owner, nameof(ICodeLensContext.GetReferenceCountAsync), new object[] { this.Descriptor, descriptorContext }, cancellationToken).ConfigureAwait(false);
+
+                if (referenceCount == null)
+                {
+                    return null;
+                }
+
+                var referenceCountString = $"{referenceCount.Count}{(referenceCount.IsCapped ? "+" : string.Empty)}";
+
+                return new CodeLensDataPointDescriptor()
+                {
+                    Description = referenceCount.Count == 1
+                        ? string.Format(CodeLensVSResources._0_reference, referenceCountString)
+                        : string.Format(CodeLensVSResources._0_references, referenceCountString),
+                    IntValue = referenceCount.Count,
+                    TooltipText = string.Format(CodeLensVSResources.This_0_has_1_references, GetCodeElementKindsString(Descriptor.Kind), referenceCountString),
+                    ImageId = null
+                };
+
+                string GetCodeElementKindsString(CodeElementKinds kind)
+                {
+                    switch (kind)
+                    {
+                        case CodeElementKinds.Method:
+                            return CodeLensVSResources.method;
+                        case CodeElementKinds.Type:
+                            return CodeLensVSResources.type;
+                        case CodeElementKinds.Property:
+                            return CodeLensVSResources.property;
+                        default:
+                            // code lens engine will catch and ignore exception
+                            // basically not showing data point
+                            throw new NotSupportedException(nameof(kind));
+                    }
+                }
+            }
+
+            public async Task<CodeLensDetailsDescriptor> GetDetailsAsync(CodeLensDescriptorContext descriptorContext, CancellationToken cancellationToken)
+            {
+                // we always get data through VS rather than Roslyn OOP directly since we want final data rather than
+                // raw data from Roslyn OOP such as razor find all reference results
+                var referenceLocationDescriptors = await _callbackService.InvokeAsync<IEnumerable<ReferenceLocationDescriptor>>(
+                    _owner, nameof(ICodeLensContext.FindReferenceLocationsAsync), new object[] { this.Descriptor, descriptorContext }, cancellationToken).ConfigureAwait(false);
+
+                var details = new CodeLensDetailsDescriptor
+                {
+                    Headers = new List<CodeLensDetailHeaderDescriptor>()
```

Comment:
Do we need to create this Headers list every time? Is it possible to have a single shared instance?